File: 112-path-sum/112-path-sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def dfs(self, root, sumTillNow):
        # print(sumTillNow, root.val)
        if not root:
            return
        if root.val + sumTillNow == self.target and root.left == root.right == None:
            self.sol = True
            return
        if root.left:
            self.dfs(root.left, sumTillNow + root.val)
        if root.right:
            self.dfs(root.right, sumTillNow + root.val)
        
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        self.target = targetSum
        self.sol = False
        if not root:
            return False
        self.dfs(root, 0)
        return self.sol

File: 112-path-sum/112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean sol = false;
    int target;
    public void dfs(TreeNode root, int sumSoFar){

        if(root == null) return;
        // System.out.println(root.val + " " + sumSoFar);
        if ((root.val + sumSoFar == this.target) && (root.left == null) && (root.right == null)) {
            this.sol = true;
            return;
        }
        if(root.left != null)
            dfs(root.left, root.val + sumSoFar);
        if (root.right != null)
            dfs(root.right, root.val + sumSoFar);
    }
    public boolean hasPathSum(TreeNode root, int targetSum) {
        this.sol = sol;
        this.target = targetSum;
        dfs(root, 0);
        return this.sol;
    }
}